### ISSUE TITLE
Associate a user to populate the AuthorOrg table

### ISSUE THREAD

**Issue by nathan-cummings-ukaea** [ScottishCovidResponse/data-registry#49]

**Comment by alahiff** [comment]
Note the ObjectAuthorOrg table has a compulsory Object associated with each entry, not just an Author and Orgs (this allows authors to be associated with different orgs for different objects).

So I don't think we can't just switch the `updated_by` fields to point to the ObjectAuthorOrg table.

**Comment by nathan-cummings-ukaea** [comment]
I don’t think using the in-built updated_by field is the correct approach, as you may be wanted to upload something that wasn’t authored by yourself.

**Comment by alahiff** [comment]
`updated_by` is defined in the `BaseModel`, so we can change it as we like.

I always assumed `updated_by` referred to the person who's actually interacting with registry (which is exactly how it is currently), rather than being associated with who actually created the object.

**Comment by alahiff** [comment]
We will create a new table (`AuthorOrg`) containing authors and associated organisations. The standard userid can then be mapped to entries in this.

**Comment by richardreeve** [comment]
The basic idea is that `updated_by` still references the userid that did the update, but that user is associated with their `AuthorOrg` entry/ies, so that we can identify exactly who updated the registry, we don't just have their userid (which may differ between local and remote registries). We could just have a `UserIDAuthorOrg` table that works just like the `ObjectAuthorOrg` but connects userids to authors and organisations. This might be simpler than having an `AuthorOrg` table and then referencing that from the other two tables (which would still be needed I think?). @alahiff does that make sense?

**Comment by alahiff** [comment]
I've created a pull request. This adds a `UserAuthorOrg` model, which associates a user with an `Author` and optional list of `Organisations`.

At the moment a user can have only a single entry in this table. The only problem with this is that if someone updates the registry, then gets a new job then does some more updates to the registry, this change in organisation can't be captured - we can only keep track of the current organisation. If we allowed multiple entries in the table for each user you still wouldn't know which changes each entry actually applies to.

**Comment by richardreeve** [comment]
@alahiff You're saying that it only gives the current organisations the person is associated with?

If so, I think that's okay. This is only for fields like `updated_by`, not authorship - am I right in thinking that if they are an author, they will have permanent entries with the correct orgs to the `ObjectAuthorOrg` table (or whatever it's called)? If so, then that's all okay I think - it just makes sure when you push to the remote registry it associates the table updates correctly with the right person... does that make sense?

**Comment by alahiff** [comment]
@richardreeve Yes, it only gives the current organisations the person is assoicated with.

Yes, if the same person is also an author they will have entries in the `ObjectAuthorOrg` table connecting them with organisations for specific objects.

Note that without the `UserAuthorOrg` table at all (i.e. what we already had before this issue was created):
* Whenever a user writes to the remote registry the `updated_by` field is updated automatically with the userid of the authenticated user. `userid` isn't a writeable field and cannot be set by users. So if data from a local registry is pushed to the remote registry the `updated_by` field in the local registry is completely ignored.
* The remote registry already has full names and lists of organisations associated with the `userid`. This information is taken from the personnel MySQL database.

**Comment by richardreeve** [comment]
Ooh, we may need to think about that down the line... it won't cause any problems for now, because we're only thinking about a single person local registry, but in principle, there's no reason why a different person may sign off on (and upload) a file that I creating in the local registry. Whether that should be marked as updated by them remotely or by me is unclear to me at the moment.

**Comment by alahiff** [comment]
@richardreeve Initially, where should the author name and org come from for the `UserAuthorOrg`? Two options I can think of are:
* Get the full name and organisation from the `User` table. In this case we could add a function so that whenever anything is updated it will automatically ensure the `UserAuthorOrg` table is up to date.
* The `fair` CLI tool could add an entry to the `UserAuthorOrg` table (so this tool would need to know the name and org(s)).

In future we can decide how to handle what happens when someone uploads files created by someone else...

**Comment by richardreeve** [comment]
I think the second option. @kzscisoft has a plan for a command line option `fair init` (I think that's the name?). This will ask for and set up a bunch of defaults, and one of those can be your name and organisation I think.

Incidentally, the reason for all of this in case it isn't clear (it doesn't seem to be mentioned in this issue!) is so that when the user creates objects using the data pipeline API, they can have an author associated with them (the user) as well as an `updated_by` entry.

**Comment by alahiff** [comment]
The second option is easier for me, so I'm happy with that :-)
